editors/code/src/client.ts

```diff
@@ -41,6 +41,7 @@ export function createClient(serverPath: string, cwd: string): lc.LanguageClient
     const clientOptions: lc.LanguageClientOptions = {
         documentSelector: [{ scheme: 'file', language: 'rust' }],
         initializationOptions: vscode.workspace.getConfiguration("rust-analyzer"),
+        diagnosticCollectionName: "rustc",
```

Comment:
Per https://github.com/microsoft/vscode-languageserver-node/blob/master/client/src/common/client.ts#L2882-L2886 this sets the name of the `owner` for the diagnostics group, which needs to match what the problem matcher produces in order for the same diagnostics to get merged.  For anyone that happens to be debugging this in the future and sees this comment, if you right click a diagnostic, select `Copy` (not `Copy Message`), you can paste the JSON representation.